main/src/main/scala/sbt/internal/InstallSbtn.scala

```diff
@@ -0,0 +1,228 @@
+    val tmp = Files.createTempFile(s"sbt-$version", "zip")
+    val sbtn = if (Properties.isWin) "sbtn.exe" else "sbtn"
+    try extractSbtn(term, version, tmp, baseDirectory.resolve("bin").resolve(sbtn))
+    finally {
+      Files.deleteIfExists(tmp)
+      ()
+    }
+    val shell = if (System.console != null) getShell(term) else "none"
+    shell match {
+      case "none" =>
+      case s =>
+        val completion = shellCompletions(s)
+        val completionLocation = baseDirectory.resolve("completions").resolve(completion)
+        downloadCompletion(completion, version, completionLocation)
+        s match {
+          case "bash"       => setupBash(baseDirectory, term)
+          case "fish"       => setupFish(baseDirectory, term)
+          case "zsh"        => setupZsh(baseDirectory, term)
+          case "powershell" => setupPowershell(baseDirectory, term)
+          case _            => // should be unreachable
+        }
+        val msg = s"Successfully installed sbtn for $s. You may need to restart $s for the " +
+          "changes to take effect."
+        term.printStream.println(msg)
+    }
+    ()
+  }
+
+  private[sbt] def extractSbtn(term: Terminal, version: String, sbtZip: Path, sbtn: Path): Unit = {
+    downloadRelease(term, version, sbtZip)
+    Files.createDirectories(sbtn.getParent)
+    val bin =
+      if (Properties.isWin) "pc-win32.exe"
+      else if (Properties.isLinux) "pc-linux"
+      else "apple-darwin"
+    val sbtnName = s"sbt/bin/sbtn-x86_64-$bin"
+    val fis = new FileInputStream(sbtZip.toFile)
+    val zipInputStream = new ZipInputStream(fis)
+    var foundBinary = false
+    try {
+      var entry = zipInputStream.getNextEntry
+      while (entry != null) {
+        if (entry.getName == sbtnName) {
+          foundBinary = true
+          term.printStream.println(s"extracting $sbtZip!$sbtnName to $sbtn")
+          transfer(zipInputStream, sbtn)
+          sbtn.toFile.setExecutable(true)
+          entry = null
+        } else {
+          entry = zipInputStream.getNextEntry
+        }
+      }
+      if (!foundBinary) throw new IllegalStateException(s"couldn't find $sbtnName in $sbtZip")
+    } finally {
+      fis.close()
+      zipInputStream.close()
+    }
+    ()
+  }
+  private[this] def downloadRelease(term: Terminal, version: String, location: Path): Unit = {
+    val zip = s"https://github.com/sbt/sbt/releases/download/v$version/sbt-$version.zip"
+    val url = new URL(zip)
+    term.printStream.println(s"downloading $zip to $location")
+    transfer(url.openStream(), location)
+  }
+  private[this] def transfer(inputStream: InputStream, path: Path): Unit = {
+    val os = new FileOutputStream(path.toFile)
+    val result = new Array[Byte](1024 * 1024)
+    var bytesRead = -1
+    try {
+      do {
+        bytesRead = inputStream.read(result)
+        if (bytesRead > 0) os.write(result, 0, bytesRead)
+      } while (bytesRead > 0)
+    } finally {
+      inputStream.close()
```

Comment:
when this throws an exception, the next line is never executed, right?

Reply:
Yes, but I have never seen an inputStream throw an exception on close.